docs/spec/governance/software_upgrade.md

```diff
@@ -0,0 +1,133 @@
+   in in the form of a special parameter.
+
+### Step 1: Agree on the Upgrade Plan
+
+First, an English plaintext proposal should be submitted based on the following template:
+
+```
+# SOFTWARE UPGRADE PROPOSAL
+
+This is a proposal to upgrade the software.
+
+## Outline:
+
+* Current version: [XXX]
+* Current version commit hash: [XXX]
+* New version Github URL: https://github.com/...
+* New version: [XXX]
+* New version commit hash: [XXX]
+* Upgrade prepared by:
+ - <name/handle> on <date>; commit signature: https://github.com/...
+* Audited by:
+ - <name/handle> on <date>; report: https://github.com/...
+ - <name/handle> on <date>; report: https://github.com/...
+
+## Description:
+
+Describe the purpose of the upgrade and other details.
+
+- [ ] Link to any previous discussions
+- [ ] Link to any alternative proposals
+
+## Requirements:
+
+Describe any requirements
+
+- [ ] Any new hardware requirements?
+- [ ] Any new bandwidth requirements?
+- [ ] Any dependencies on other proposals?
+
+## Procedure:
+
+* Deadline: By 2 weeks from after the proposal has been accepted.
+
+Once this proposal has been accepted, validators should upgrade to the new
+software as described above by the deadline.  Once the new software is
+deployed, the new software will automaticaly submit a message to the
+SoftwareUpgradeKeeper.  The SoftwareUpgradeKeeper ensures that a sufficient
+quorum of validators (SOFTWARE_UPGRADE_QUORUM parameter) is running a specific
+version of the blockchain software.
+
+Once a sufficient quorum has been reached, the "SOFTWARE_UPGRADE_[XXX]_ENABLED"
+parameter will get set to true, thus switching on the new logic.
+
+## Conflict Resolution
+
+In case there are conflicting proposals, the most recent proposal to get
+accepted should be followed.  In other words, if another proposal gets accepted
+after this proposal does which conflicts with this proposal, the new proposal
+should be heeded instead of the process declared here.
+```
+
+It is recommended that all links point to Github for now, for accountability.
+
+This proposal template assumes that much discussion about the upgrade has
+already taken place elsewhere.  To gauge interest in a proposed upgrade,
+users should instead submit a survey proposal with the following template:
+
+Notethat the `Procedure` section has already been filled out.  Do not change
+this section unless there is a good reason to do so.
+
+```
+# SOFTWARE UPGRADE SURVEY
+
+THIS IS A NONBINDING SURVEY.  VOTES HERE ARE ONLY INTENDED TO GUAGE INTEREST.
+IF THIS PROPOSAL MANDATES ANY ACTION, IT IS THE VOTER'S DUTY TO VOTE "NO".
+
+<fill out as much as possible from the SOFTWARE UPGRADE PROPOSAL template>
+```
+
+### Step 2: Validators upgrade their software
+
+Once the proposal from step 1 has been accepted, validators should upgrade
+their software.  For the standard upgrade procedure, it is OK for non-validator
+full nodes to upgrade as well.
```

Comment:
isn't it required that full nodes upgrade as well? - I mean if they don't upgrade they just risk not being able to run transactions

Reply:
Depends on the upgrade (e.g. with an upgrade which strictly reduces the set of valid transactions, they don't really need to upgrade).

In most cases they'll need to, the node should warn the user.